Draft for PR-3088 of Issue-637

stt08 opened the pull request.

stt08 committed:
formatting changes, removal of unnecessary functions

JuditKnoll commented on CHANGELOG.md:
```diff
@@ -1,11 +1,12 @@
-
```
Do not change the changelog unnecessaryly. Especially the older, already released changes. Most likely these came in when merging.

JuditKnoll replied:
There are still a lot of changes, but now, these seems to be only whitespace changes. Can you please fix them?

JuditKnoll commented on spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue637Test.java:
```diff
@@ -0,0 +1,89 @@
+ * SpotBugs should check for bad combinations of date format flags in SimpleDateFormat.
+ *
+ * @see <a href="https://github.com/spotbugs/spotbugs/issues/637">GitHub issue</a>
+ * @since 4.4.2
```
The `@since` part is unnecessary. Can you please delete it?

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,146 @@
+        }
+
+        String sig = getSigConstantOperand();
+        boolean hasProperSignal = sig.contains("Ljava/lang/String;)V") || sig.contains("Ljava/lang/String;Ljava/util/Locale;)V")
```
Here `sig` stands for signature, and not for signal. Can you please rename it?

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,146 @@
+        String sig = getSigConstantOperand();
+        boolean hasProperSignal = sig.contains("Ljava/lang/String;)V") || sig.contains("Ljava/lang/String;Ljava/util/Locale;)V")
+                || sig.contains("Ljava/lang/String;Ljava/text/DateFormatSymbols;)V");
+        boolean isProperFunction = Arrays.asList("<init>", "applyPattern", "applyLocalizedPattern")
```
There is a constant for the `"<init>"` string: `Const.CONSTRUCTOR_NAME`. Can you please use that instead?

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,146 @@
+        }
+        if (dateFormatString == null) {
+            return;
+        }
```
This can be moved just before the first use of the string. Also, the relevant string should be consistently one element for one function call. Can you please use the relevant one every time?

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,146 @@
+        boolean hasProperSignal = sig.contains("Ljava/lang/String;)V") || sig.contains("Ljava/lang/String;Ljava/util/Locale;)V")
+                || sig.contains("Ljava/lang/String;Ljava/text/DateFormatSymbols;)V");
+        boolean isProperFunction = Arrays.asList("<init>", "applyPattern", "applyLocalizedPattern")
+                .contains(getNameConstantOperand());
```
These variables are only used once. Are they really necessary?

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,146 @@
+        boolean isProperFunction = Arrays.asList("<init>", "applyPattern", "applyLocalizedPattern")
+                .contains(getNameConstantOperand());
+
+        if (hasProperSignal && isProperFunction && getClassConstantOperand().equals("java/text/SimpleDateFormat")) {
```
Is there any relevant class other than `SimpleDateFormat`? Are there other relevant functions?

stt08 replied:
Added more classes and methods

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,146 @@
+     */
+    private boolean runDateFormatRuleVerify(String dateFormat) {
+        // standard time (x2), military time (x2), week-year flag (x1)
+        DateFormatRule[] badCombinations = {
+            new DateFormatRule("a", "h"), new DateFormatRule("a", "K"),
+            new DateFormatRule(null, "H", "a"), new DateFormatRule(null, "k", "a"),
+            new DateFormatRule("w", "Y", "M", "d") };
```
Are there any other bad combinations?

stt08 replied:
I dont think that there is anything else to be added. Those mentioned in the PR were already implemented

JuditKnoll replied:
For [DateTimeFormatter](https://docs.oracle.com/en%2Fjava%2Fjavase%2F22%2Fdocs%2Fapi%2F%2F/java.base/java/time/format/DateTimeFormatter.html) there are several new pattern symbols introduced. Did you check those as well?

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,146 @@
+     * Runs the check per each bad combination (DateFormatRule objects) on the provided string.
+     *
+     * @param dateFormat - string to be checked
+     * @return {@code true} if given string any bad combination matches.
```
```suggestion
     * @return {@code true} if given string matches any bad combination.
```

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,146 @@
+        this.bugReporter = bugReporter;
+    }
+
+    /**
+     * Checks whether the current element is a proper function (one of three specified)
+     * that has at least 1 parameter. Such methods would have 1st parameter to appear as
+     * dateFormat string, which will be tested for forbidden combinations.
+     */
```
Are you sure this javadoc describes well what the function does? Please, either rewrite it or remove it.

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,146 @@
+
+/**
+ * This detector can find ...
+ */
```
This javadoc comment is unfinished. Please, either finish it or remove it.

renovate[bot] committed:
fix(deps): update dependency org.mockito:mockito-core to v5.13.0 (#3090)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency com.google.errorprone:error_prone_annotations to v2.31.0 (#3091)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

stt08 committed:
removal of variables, changelog restored

gtoison committed:
RocksDB ReadOptions not recognized as closeable (#3076)

* test: issue #3069 reproducer

* fix: added a built-in obligation database
* fix: added obligations database for ReadOptions' constructors and close

renovate[bot] committed:
fix(deps): update dependency org.apache.ant:ant to v1.10.15 (#3093)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency org.apache.commons:commons-lang3 to v3.17.0 (#3096)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update plugin com.github.spotbugs to v6.0.21 (#3097)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency checkstyle to v10.18.1 (#3098)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,129 @@
+                && getSigConstantOperand().contains("Ljava/lang/String;")
+                && "java/text/SimpleDateFormat".equals(getClassConstantOperand())) {
+
+            String dateFormatString = (String) stack.getStackItem(1).getConstant();
+            if (new SignatureParser(getSigConstantOperand()).getNumParameters() == 1) {
+                dateFormatString = (String) stack.getStackItem(0).getConstant();
+            }
```
This part is not so easy to understand why it's like this. Can you please check the exact signatures (or at least the parameter parts) for the different functions and set this `dateFormatString` variable according to that?

stt08 replied:
fixed: it uses position of latest parameter in the stack (as the order is reversed)

JuditKnoll replied:
It's still not easy to get why that is the relevant parameter. That is only a lucky coincidence, that it's every time that parameter. It can be different for different classes and functions. Can you please check the exact signatures (or at least the parameter parts) for the different functions and set this `dateFormatString` variable according to that?

stt08 replied:
fixed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,129 @@
+                dateFormatString = (String) stack.getStackItem(0).getConstant();
+            }
+
+            if(dateFormatString != null && runDateFormatRuleVerify(dateFormatString)) {
```
As far as I can see, `dateFormatString` can never be null here, so the null check is not necessary.

stt08 replied:
There are some cases where it is possible to receive a null. It used to be the case in applyLocalizedPattern method when String was stored inside the variable to which localisation was applied first. Currently it uses String value instead

JuditKnoll replied:
Thanks for the explanation :)

JuditKnoll commented on CHANGELOG.md:
```diff
@@ -1,11 +1,12 @@
 Currently the versioning policy of this project follows [Semantic Versioning v2.0.0](http://semver.org/spec/v2.0.0.html).
 
 ## Unreleased - 2024-??-??
+### Added
+- Add SimpleDateFormat string check for bad combinations of format flags ([#637](https://github.com/spotbugs/spotbugs/issues/637))
 ### Fixed
```
Can you please add an empty line before ###Fixed?

stt08 replied:
fixed

JuditKnoll reviewed (changes requested):
The build is failing because of some format issues. Please fix them by running `gradlew spotlessApply`. Before asking for a review, please make sure that `gradlew build smokeTest` is successful.

stt08 committed:
index of dateFormatString calculated, changelog restored

JuditKnoll committed:
Partial fix for Issue #3023 (#3080)

* issue reproducer
* fix for classes
* disable still breaking test
* add changelog entry
* add test with both mutable and immutable assignments
* fix FNs with ternary operator

JuditKnoll committed:
small fixes according to sonar cloud (#3092)

JuditKnoll commented on CHANGELOG.md:
```diff
@@ -16,6 +19,7 @@ Currently the versioning policy of this project follows [Semantic Versioning v2.
 - Revert the changes from ([#2894](https://github.com/spotbugs/spotbugs/pull/2894)) to get HTML stylesheets to work again ([#2969](https://github.com/spotbugs/spotbugs/issues/2969))
 - Fix FP `SING_SINGLETON_GETTER_NOT_SYNCHRONIZED` report when the synchronization is in a called method ([#3045](https://github.com/spotbugs/spotbugs/issues/3045))
 - Let `BetterCFGBuilder2.isPEI` handle `dup2` bytecode used by Spring AOT ([#3059](https://github.com/spotbugs/spotbugs/issues/3059))
+- Detect failure to close RocksDB's ReadOptions ([#3069](https://github.com/spotbugs/spotbugs/issues/3069))
```
This was already added to the upstream repository. If you sync or update your master, this will not be marked as new.

stt08 replied:
fixed

JuditKnoll commented on spotbugsTestCases/src/java/ghIssues/issue637/Issue637NoErrors.java:
```diff
@@ -0,0 +1,119 @@
+        SimpleDateFormat hour24format2 = new SimpleDateFormat(MILITARY_HOUR_1_24_CORRECT);
+        SimpleDateFormat weekYear = new SimpleDateFormat(WEEK_YEAR_CORRECT);
+
+        String localizedHour12format1 = hour12format1.toLocalizedPattern();
+        String localizedHour12format2 = hour12format2.toLocalizedPattern();
+        String localizedHour24format1 = hour24format1.toLocalizedPattern();
+        String localizedHour24format2 = hour24format2.toLocalizedPattern();
+        String localizedWeekYear = weekYear.toLocalizedPattern();
```
Can you please remove these `toLocalizedPattern()` calls or add a test without them to show, these doesn't hide FPs?

stt08 replied:
fixed

gtoison committed:
Fixes for Kotlin `NotNull` (#3095)

* test: reproducer for issue #3094

* fix: ignore Kotlin generated constructors in ...

... InitializeNonnullFieldsInConstructor
fix: recognize Kotlin's Intrinsics.checkNotNullParameter() as a nullness
checker

* Added Kotlin Intrinsics.checkXyz methods to the database

Added the code sample submitted by the issue reporter for future
reference

* fix: applied spotless

renovate[bot] committed:
fix(deps): update dependency org.checkerframework:checker-qual to v3.47.0 (#3101)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update plugin com.github.spotbugs to v6.0.22 (#3103)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

stt08 committed:
code placement change

stt08 committed:
code placement change

stt08 committed:
Merge branch 'master' into issue-637

stt08 committed:
Merge branch 'master' of github.com:stt08/spotbugs into issue-637

stt08 committed:
conflict of changelog

stt08 committed:
conflict of changelog

stt08 committed:
conflict resolve

renovate[bot] committed:
fix(deps): update log4j2 monorepo to v2.24.0 (#3104)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency gradle to v8.10.1 (#3105)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

stt08 committed:
more classes added to be checked

renovate[bot] committed:
chore(deps): update plugin com.gradle.develocity to v3.18.1 (#3107)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency com.google.errorprone:error_prone_annotations to v2.32.0 (#3108)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency org.apache.groovy:groovy-all to v4.0.23 (#3109)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency joda-time:joda-time to v2.13.0 (#3112)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency com.diffplug.gradle:goomph to v4.1.0 (#3116)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

stt08 committed:
code structure and logic changed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+     * @return {@code true} if given string matches any bad combination.
+     */
+    private boolean runDateFormatRuleVerify(String dateFormat) {
+        for (Rule rule : new Rule[] {
+            // when "h" or "K" flags found, make sure that it ALSO CONTAINS "a" or "B"
+            new Rule(Arrays.asList("a", "B"), null, true, Arrays.asList("h", "K")),
+
+            // when "H" or "k" flags found, make sure that it DOES NOT CONTAIN "a" or "B"
+            new Rule(Arrays.asList("a", "B"), null, false, Arrays.asList("H", "k")),
+
+            // when "M" or "d" flags found, make sure that it DOES NOT CONTAIN "Y" (unless "w" is provided)
+            new Rule(Collections.singletonList("Y"), "w", false, Arrays.asList("M", "d")),
+
+            // milli-of-day cannot be used together with Hours, Minutes, Seconds
+            new Rule(Arrays.asList("H", "h", "K", "k", "m", "s"), null, false, Arrays.asList("A", "N")),
+
+            // milli-of-day and nano-of-day cannot be used together
+            new Rule(Collections.singletonList("A"), null, false, Collections.singletonList("N")),
+            new Rule(Collections.singletonList("N"), null, false, Collections.singletonList("A")),
+
+            // fraction-of-second and nano-of-second cannot be used together
+            new Rule(Collections.singletonList("S"), null, false, Collections.singletonList("n")),
+            new Rule(Collections.singletonList("n"), null, false, Collections.singletonList("S")),
+
+            // am-pm marker and period-of-day cannot be used together
+            new Rule(Collections.singletonList("a"), null, false, Collections.singletonList("B")),
+            new Rule(Collections.singletonList("B"), null, false, Collections.singletonList("a")),
+
+            // year and year-of-era cannot be used together
+            new Rule(Collections.singletonList("u"), null, false, Collections.singletonList("y")),
+            new Rule(Collections.singletonList("y"), null, false, Collections.singletonList("u"))
+        }) {
```
Please do not create the array in the for cycle's header. This makes the code hard to read. However, I really like the comments, which makes the rules easy to understand.

stt08 replied:
fixed - rewriten into stream

JuditKnoll commented on spotbugs/etc/messages.xml:
```diff
@@ -7502,6 +7510,21 @@ This format string includes a newline character (\n). In format strings, it is g
+    <Details>
+      <![CDATA[
+      <p>
+      This format string includes a bad combination of flags which may lead to unexpected behavior.
+      Potential bad combinations include the following: using a week year ("Y") with month in year
+      ("M") and day in month ("d") without specifying week in year ("w") ("y" may be preferable
+      here instead); using an AM/PM hour ("h" or "K") without specifying an AM/PM marker ("a");
+      and using a 24-hour format hour ("H" or "k") and also specifying an AM/PM marker.
```
Please make sure, that the details section contains every problematic flag combination.

stt08 replied:
fixed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+
+    /**
+     * Contains special flags that can trigger check (keywords); property of the rule
+     * (isRequired - whether flag should contain or be absent); flags to be checked
```
```suggestion
     * (isRequired - whether flag should be contained or be absent); flags to be checked
```

stt08 replied:
fixed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+     * and a bypassing flag (whether a rule can be skipped if such flag is included)
+     */
+    private static final class Rule {
+        boolean isRequired;
+        String ignore;
+        List<String> flags;
+        List<String> keywords;
```
As far as I can see, these fields could be private final.

stt08 replied:
fixed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+
+        String dateFormatString = (String) stack.getStackItem(idx).getConstant();
+        if (dateFormatString != null && runDateFormatRuleVerify(dateFormatString)) {
+            bugReporter.reportBug(new BugInstance(this, "FS_BAD_DATE_FORMAT_FLAG_COMBO", NORMAL_PRIORITY)
+                    .addClassAndMethod(this).addCalledMethod(this).addString(dateFormatString)
+                    .describe(StringAnnotation.FORMAT_STRING_ROLE).addSourceLine(this));
```
```suggestion
            bugReporter.reportBug(new BugInstance(this, "FS_BAD_DATE_FORMAT_FLAG_COMBO", NORMAL_PRIORITY)
                    .addClassAndMethod(this)
                    .addCalledMethod(this)
                    .addString(dateFormatString).describe(StringAnnotation.FORMAT_STRING_ROLE)
                    .addSourceLine(this));
```
Can you please reformat it similarly so it's easy to see what belongs together?

stt08 replied:
fixed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+        }
+    }
+
+    private boolean contains(String item, String... items) {
+        return Arrays.asList(items).contains(item);
+    }
```
This function is quite simple and only used once. Does it help to have it in a separate function?

stt08 replied:
fixed - removed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+    private static final class Rule {
+        boolean isRequired;
+        String ignore;
+        List<String> flags;
+        List<String> keywords;
```
What's the difference between `flags` and `keywords`? Can you please give more descriptive names?

stt08 replied:
fixed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+            this.keywords = keywords;
+        }
+
+        boolean containsAny(String target, List<String> list) {
```
Can you please give more descriptive names to the parameters?

stt08 replied:
fixed + comment above

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+        }
+
+        boolean containsAny(String target, List<String> list) {
+            if (list == null) {
+                return false;
+            }
+            for (String item : list) {
+                if (target.contains(item)) {
+                    return true;
+                }
+            }
+            return false;
```
This may be a bit more elegant with streams like this:
```java
return list != null && list.stream().anyMatch(target::contains);
```
What do you think?

stt08 replied:
Perfect, it is being used now

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+     * @return {@code true} if given string matches any bad combination.
+     */
+    private boolean runDateFormatRuleVerify(String dateFormat) {
+        for (Rule rule : new Rule[] {
+            // when "h" or "K" flags found, make sure that it ALSO CONTAINS "a" or "B"
+            new Rule(Arrays.asList("a", "B"), null, true, Arrays.asList("h", "K")),
+
+            // when "H" or "k" flags found, make sure that it DOES NOT CONTAIN "a" or "B"
+            new Rule(Arrays.asList("a", "B"), null, false, Arrays.asList("H", "k")),
+
+            // when "M" or "d" flags found, make sure that it DOES NOT CONTAIN "Y" (unless "w" is provided)
+            new Rule(Collections.singletonList("Y"), "w", false, Arrays.asList("M", "d")),
+
+            // milli-of-day cannot be used together with Hours, Minutes, Seconds
+            new Rule(Arrays.asList("H", "h", "K", "k", "m", "s"), null, false, Arrays.asList("A", "N")),
+
+            // milli-of-day and nano-of-day cannot be used together
+            new Rule(Collections.singletonList("A"), null, false, Collections.singletonList("N")),
+            new Rule(Collections.singletonList("N"), null, false, Collections.singletonList("A")),
+
+            // fraction-of-second and nano-of-second cannot be used together
+            new Rule(Collections.singletonList("S"), null, false, Collections.singletonList("n")),
+            new Rule(Collections.singletonList("n"), null, false, Collections.singletonList("S")),
+
+            // am-pm marker and period-of-day cannot be used together
+            new Rule(Collections.singletonList("a"), null, false, Collections.singletonList("B")),
+            new Rule(Collections.singletonList("B"), null, false, Collections.singletonList("a")),
+
+            // year and year-of-era cannot be used together
+            new Rule(Collections.singletonList("u"), null, false, Collections.singletonList("y")),
+            new Rule(Collections.singletonList("y"), null, false, Collections.singletonList("u"))
+        }) {
+            if (rule.verify(dateFormat)) {
+                return true;
+            }
+        }
+        return false;
```
The usage of streams could result in a more elegant code here as well.

stt08 replied:
fixed

JuditKnoll commented on spotbugs/etc/messages.xml:
```diff
@@ -7502,6 +7510,21 @@ This format string includes a newline character (\n). In format strings, it is g
   </BugPattern>
+  <BugPattern type="FS_BAD_DATE_FORMAT_FLAG_COMBO">
+    <ShortDescription>Certain SimpleDateFormat format strings may lead to unexpected behavior</ShortDescription>
+    <LongDescription>Certain SimpleDateFormat format strings may lead to unexpected behavior in {1}</LongDescription>
```
Instead of the word `Certain` the exact problematic string could be referenced here. It's already added to the BugInstance.

stt08 replied:
fixed

JuditKnoll commented on spotbugs/etc/messages.xml:
```diff
@@ -7502,6 +7510,21 @@ This format string includes a newline character (\n). In format strings, it is g
      </Details>
   </BugPattern>
+  <BugPattern type="FS_BAD_DATE_FORMAT_FLAG_COMBO">
+    <ShortDescription>Certain SimpleDateFormat format strings may lead to unexpected behavior</ShortDescription>
+    <LongDescription>Certain SimpleDateFormat format strings may lead to unexpected behavior in {1}</LongDescription>
```
Now, it not only covers `SimpleDateFormat`, but other classes as well, so in certain cases this bug description could be a bit confusing.

stt08 replied:
fixed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+        String nm = getNameConstantOperand();
+        String si = getSigConstantOperand();
+        if ("java/text/SimpleDateFormat".equals(cl)
+                && contains(nm, Const.CONSTRUCTOR_NAME, "applyPattern", "applyLocalizedPattern")) {
+            switch (si) {
+            case "(Ljava/lang/String;)V":
+                idx = 0;
+                break;
+            case "(Ljava/lang/String;Ljava/util/Locale;)V":
+            case "(Ljava/lang/String;Ljava/text/DateFormatSymbols;)V":
+                idx = 1;
+                break;
+            }
```
Can all of the cases happen with every function?

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+                break;
+            }
+        } else if ("java/time/format/DateTimeFormatter".equals(cl) && "ofPattern".equals(nm)) {
+            switch (si) {
```
Can you please provide a few lines of comments to the switches, why that is the value of `idx`?

stt08 replied:
fixed

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,183 @@
+                break;
+            }
+        } else if ("java/time/format/DateTimeFormatter".equals(cl) && "ofPattern".equals(nm)) {
+            switch (si) {
+            case "(Ljava/lang/String;)Ljava/time/format/DateTimeFormatter;":
+                idx = 0;
+                break;
+            case "(Ljava/lang/String;Ljava/util/Locale;)Ljava/time/format/DateTimeFormatter;":
+                idx = 1;
+                break;
+            }
```
Switch statements are useful when there are many different cases depending on the value of the same expression.
For just one or two cases, the code is more readable with if statements.

stt08 replied:
fixed

gtoison committed:
ci: run the build against JDK 23 (#3117)

renovate[bot] committed:
fix(deps): update dependency com.google.guava:guava to v33.3.1-jre (#3119)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency gradle to v8.10.2 (#3118)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update plugin com.github.spotbugs to v6.0.23 (#3121)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update junit5 monorepo to v5.11.1 (#3122)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency org.mockito:mockito-core to v5.14.0 (#3126)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

JuditKnoll committed:
Add bug checking assert functions to `AbstractIntegrationTest` (#3120)

gtoison committed:
CWE mapping fixes (#3124)

* fix: CWE mapping fixes

* fix: mapped FI_MISSING_SUPER_CALL to CWE 568

gtoison committed:
Extends `@SuppressFBWarnings` to allow regex and exact matches (#3102)

Extend the bug suppression, allowing exact (not prefix) bug type match,
and regex bug type match

renovate[bot] committed:
fix(deps): update log4j2 monorepo to v2.24.1 (#3127)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency checkstyle to v10.18.2 (#3128)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

hazendaz committed:
Update gradle and maven plugin versions that work with 4.8.6 (#3129)

hazendaz committed:
Cleanup regex and thread issue in test harness (#3130)

* Cleanup regex and thread issue in test harness

* update changelog

hazendaz committed:
Remove extra empty space / Remove redundent publics from interfaces (#3131)

hazendaz committed:
Fix order of modifiers on properties / methods and ensure correct location in file (#3132)

x

hazendaz committed:
Return objects directly instead of creating more GCV by defining them (#3133)

* Return objects directly instead of creating more GCV by defining them

* fix, missed item on comment

stt08 committed:
improvement of code, description change

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,187 @@
+            bugReporter.reportBug(new BugInstance(this, "FS_BAD_DATE_FORMAT_FLAG_COMBO", NORMAL_PRIORITY)
+                    .addClassAndMethod(this)
+                    .addCalledMethod(this)
+                    .addString(cl).addString(dateFormatString).describe(StringAnnotation.FORMAT_STRING_ROLE)
```
`cl` is the class of the called method. Please only use `addString()` when there are no more specific method (e.g. `addMethod()` or `addClass()`) to add the information.

stt08 replied:
fixed

JuditKnoll commented on spotbugs/etc/messages.xml:
```diff
@@ -7502,6 +7510,25 @@ This format string includes a newline character (\n). In format strings, it is g
+      period of day markers; using a milli of day ("A") together with hours ("H", "h", "K", "k")
+      and/or minutes ("m") and/or seconds ("s"); use of milli of day ("A") and nano of day ("N")
+      together; use of fraction of second ("S") nano of second together ("n"); use of AM/PM markers
+      ("a") and period of day ("B") together and use of year ("y") and year of era ("u") together.
```
IMO it would be a bit easier to read if the combinations would be in a list.

stt08 replied:
fixed

hazendaz committed:
Cleanup double initialization (#3134)

stt08 committed:
changelog modifications

stt08 committed:
Merge branch 'master' into issue-637

stt08 committed:
call of tests using provided methods

stt08 committed:
bad combinations in messages are listed as list

stt08 committed:
spotless apply

stt08 committed:
messages are updated with proper formatting

JuditKnoll commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@ Currently the versioning policy of this project follows [Semantic Versioning v2.
 ## Unreleased - 2024-??-??
 ### Added
 - Updated the `SuppressFBWarnings` annotation to support finer grained bug suppressions ([#3102](https://github.com/spotbugs/spotbugs/pull/3102))
+- Add SimpleDateFormat string check for bad combinations of format flags ([#637](https://github.com/spotbugs/spotbugs/issues/637))
```
Right now, the detector also handles other classes as well, but this changelog entry only refers to SimpleDateFormat. Can you please update it?

JuditKnoll commented on spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java:
```diff
@@ -0,0 +1,174 @@
+               -1 when no method found, or has different number of parameters
+            within the specified functions (given in switch and if) */
+        if (("java/text/SimpleDateFormat".equals(cl) && Arrays.asList(Const.CONSTRUCTOR_NAME,
+                "applyPattern", "applyLocalizedPattern").contains(nm)
```
I'm not so fond of this `Arrays.asList(item1, item2).contains(sg)` pattern, as it is not really efficient. Also, it makes it easier to read when one part of a condition is in one line, if it is possible (maybe it was in one line earlier, but spotless modified it).

JuditKnoll reviewed (changes requested):
Two small nitpicking, otherwise it's good to go :)

stt08 committed:
or-checks instead of array.contains and changelog modificaitons
